### PR TITLE
Port graphics backend to WGPU

### PR DESCRIPTION
**Pull request by hasenbanck**
This is  a big one. The main reason for me porting the backend is the included support for all the three native, new graphic APIs (DX12, Metal and Vulkan). We don't use any "advance" features like ray tracing or mesh shaders, so the transformation was pretty "lossless" in terms of features. On my system the performance on Vulkan was comparable, but we use Vulkan pretty "OpenGL"-like right now and are CPU limited. Once we go bindless, any performance differences should be negletable.

WGPU is a bit more verbose, so in the short term the code in the sub-renderer look a bit bloated, but once we use bindless and reduce the amount of bind groups (descriptor sets in Vulkan) and pipelines, things should start to look much more tidy.

Some other changes I had to do:
 - Rework anc leaned up the camera system. I now properly defined our coordinate system and cleaned up most irregularities (I think). I tried to make clear when we use world space, clip space or screen space. There are still some rough edges left though.
 - We now use infinite, reverse-Z projection. I needed to rework the depth offset of entities and it's shadows. Z-fighting shouldn't occure anymore.
 - Using stable WGPU required us to update to winit 0.30. There are two deprecated functions, but I didn't want to refactor them too, since that would have bloated the PR even more.
 - The way we render the GUI and WGPU's lifetime requirements on renderpasses requires that we pass them through the interface code. Maybe the interface reqwrite could alleviate this pain (by producing render instructions, so that we decouple the interface and the graphics backend code more).
 - All shaders now use WGSL. I cleaned up some rough edges here and there.
 - Buffer and Push constants need to be defined in WGSL and Rust. There is sadly no abstraction that Vulkano provided.
 - I added a prototype version on how we could make the graphics backend use optional "features", in this case partial buffer bindings.
 - Some helper / wrapper functionality is provided (for example attachment creation and buffer handling).

Open issues

- [x] The shadows are "darker" in this branch. I will need to compare the output of the old and new version to find the root cause.
- [X] Artifacts seems to be visible under AMD radv linux driver.
- [x] Point Lights don't work right now.
- [x] Under Linux (wayland?) keyboard input inside the chat is registered two times.
- [ ] Currently shadows of entities don't look as nice. Maybe you could have a look at the current depth calculation for entities and give me some feedback. Curveture isn't used at all right now.

Another benefit of this PR: Our graphics backend uses no unsafe code anymore.

### CODE REVIEW

**Review by vE5li — COMMENTED** [review]
Incredible work, I can't wait to merge :)

File: korangar/src/graphics/renderers/shadow/indicator/mod.rs [hunk 1]
```diff
@@ -1,100 +1,165 @@
...
             view_projection: (projection_matrix * view_matrix).into(),
-            upper_left: Padded(upper_left.into()),
-            upper_right: Padded(upper_right.into()),
-            lower_left: Padded(lower_left.into()),
-            lower_right: lower_right.into(),
+            upper_left: upper_left.extend(1.0).into(),
+            upper_right: upper_right.extend(1.0).into(),
+            lower_left: lower_left.extend(1.0).into(),
+            lower_right: lower_right.extend(1.0).into(),
         };
```
[the remark this answers is not in the document]
Comment (hasenbanck) [reply korangar/src/graphics/renderers/shadow/indicator/mod.rs:159]: It all comes down to memory layout constraints. std430 memory layouts (the standard for push constants in Vulkan) needs vec3 to be padded to vec4. This translates to other backends as far as I know. There is the optional scalar layout for Vulkan, but it's not portable to the other backends.

File: korangar/src/graphics/renderers/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /// Convince function to create [`BindingResource`] from the buffer.
+    pub fn as_entire_binding(&self) -> BindingResource<'_> {
+        self.buffer.as_entire_binding()
+    }
+
+    /// Used when the user wants to write data into a buffer.
+    /// Will re-allocate the buffer is it is too small.
+    pub fn write(&mut self, device: &Device, queue: &Queue, data: &[T]) {
```
[the remark this answers is not in the document]
Comment (hasenbanck) [reply korangar/src/graphics/renderers/buffer.rs:81]: Let's remove it for now. I will save the current branch and can re-add it later, once I rework the render to use a bindless architecture (where we upload all render instructions als buffer and the shader can go brrrrr).

File: korangar/src/graphics/renderers/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            Bound::Excluded(&n) => (n * size_of::<T>()) as u64,
+            Bound::Unbounded => self.size.load(Ordering::Acquire),
+        };
+
+        self.buffer.slice(start..end)
+    }
+
+    /// Returns the current size of data in bytes that is written inside the
+    /// underlying GPU buffer.
+    pub fn byte_size(&self) -> Option<BufferSize> {
```
[the remark this answers is not in the document]
Comment (hasenbanck) [reply korangar/src/graphics/renderers/buffer.rs:156]: Dito as above.

File: korangar/src/graphics/renderers/deferred/effect/effect.wgsl [hunk 4]
```diff
@@ -0,0 +1,60 @@
+struct Constants {
+    top_left: vec2<f32>,
+    bottom_left: vec2<f32>,
+    top_right: vec2<f32>,
+    bottom_right: vec2<f32>,
+    t_top_left: vec2<f32>,
+    t_bottom_left: vec2<f32>,
+    t_top_right: vec2<f32>,
+    t_bottom_right: vec2<f32>,
```
[the remark this answers is not in the document]
Comment (hasenbanck) [reply korangar/src/graphics/renderers/deferred/effect/effect.wgsl:9]: We can just call it `texture_` I guess!

**Pull request merged**